File: server_1/www/tests/js/controlCode.js
let currentResponse = false;

function luhnValidation(code) {
  const numbers = Array.from(String(code), Number).reverse();
  let total = 0;
  for (let key = 0; key < numbers.length; key++) {
    let number = numbers[key];
    if (key % 2 !== 0) number *= 2;
    total += number > 9 ? number - 9 : number;
  }
  return total % 10 === 0 ? true : false;
}

// Fonction obligatoire
function stdin() {
    const code = Array.from({ length: 15 }, () => Math.floor(Math.random() * 9) + 1).join('');
    currentResponse = luhnValidation(code);
    return code;
}

// Fonction obligatoire
function hoTryCode() {
  if (typeof main === 'function') {
    try {
      const returnBoolean = main();
      if (returnBoolean === currentResponse) {
        return true;
      } else {
        return "Votre code ne correspond pas à l'algorithme attendu.";
      }
    } catch (error) {
      return "Il y a une erreur dans votre code : " + error.message;
    }
  } else {
    return 'Erreur : fonction main introuvable.';
  }
}

File: .docker/server_http/www/tests/python/controlCode.py
import random

currentResponse = False

def luhnValidation(code):
    numbers = list(map(int, str(code)))[::-1]
    total = 0
    for key, number in enumerate(numbers):
        if key % 2 != 0:
            number *= 2
        total += number if number <= 9 else number - 9
    return total % 10 == 0

def stdin():
    global currentResponse
    code = ''.join(str(random.randint(1, 9)) for _ in range(15))
    currentResponse = luhnValidation(code)
    return code

def hoTryCode():
    global currentResponse
    if 'main' in globals() and callable(main):
        try:
            returnBoolean = main()
            if returnBoolean == currentResponse:
                print(True)
            else:
                print("Votre code ne correspond pas à l'algorithme attendu.")
        except Exception as error:
            print("Il y a une erreur dans votre code : " + str(error))
    else:
        print('Erreur : fonction main introuvable.')